include/swift/SILOptimizer/Utils/InstOptUtils.h

```diff
@@ -373,48 +203,92 @@ bool tryCheckedCastBrJumpThreading(
 
 /// A utility for deleting one or more instructions belonging to a function, and
 /// cleaning up any dead code resulting from deleting those instructions. Use
-/// this utility instead of
-/// \c recursivelyDeleteTriviallyDeadInstruction.
+/// this utility instead of \p recursivelyDeleteTriviallyDeadInstruction
+/// as follows:
+///   InstructionDeleter deleter;
+///   deleter.deleteIfDead(instruction);
+///   deleter.cleanupDeadInstructions();
+///
+/// This is designed to be used with a single 'onDelete' callback, which is
+/// invoked consistently just before deleting each instruction. It's usually
+/// used to avoid iterator invalidation (see the updatingIterator() factory
+/// method). The other InstModCallbacks should generally be handled at a higher
+/// level, and avoided altogether if possible. The following two are supported
+/// for flexibility:
+///
+/// callbacks.createdNewInst() is invoked incrementally when it fixes lifetimes
+/// while deleting a set of instructions, but the SIL may still be invalid
+/// relative to the new instruction.
+///
+/// callbacks.notifyWillBeDeletedFunc() is invoked when a dead instruction is
+/// first recognized and was not already passed in by the client. During the
+/// callback, the to-be-deleted instruction has valid SIL. It's operands and
+/// uses can be inspected and cached. It will be deleted later during
+/// cleanupDeadInstructions().
+///
+/// Note that the forceDelete* APIs only invoke notifyWillBeDeletedFunc() when
+/// an operand's definition will become dead after force-deleting the specified
+/// instruction. Some clients force-delete related instructions one at a
+/// time. It is the client's responsiblity to invoke notifyWillBeDeletedFunc()
+/// on those explicitly deleted instructions if needed.
+///
+/// TODO: Move this into a separate header.
```

Comment:
I think it makes sense to do this now instead of adding ` UpdatingInstructionIterator.h` - and put the logic of `UpdatingInstructionIteratorRegistry` directly into `InstructionDeleter` (see my other comment).

Reply:
I moved InstModCallbacks and UpdatingInstructionIterator because those classes completely obscured the important InstOptUtils API declarations. Those two classes are implementation details that people never need to look at unless they care about those implementations.

InstructionDeleter is a top-level API that people need to scan all the time. I don't really care if InstructionDeleter goes in a separate header, but that's not necessary and it would still need to be referenced from InstOptUtils.h.

Reply:
ok